src/client/ua_client_connect.c

```diff
@@ -1513,6 +1538,141 @@ UA_Client_connectSecureChannel(UA_Client *client, const char *endpointUrl) {
     return client->connectStatus;
 }
 
+static void
+__Client_reverseConnectCallback(UA_ConnectionManager *cm, uintptr_t connectionId,
+                         void *application, void **connectionContext,
+                         UA_ConnectionState state, const UA_KeyValueMap *params,
+                         UA_ByteString msg) {
+    void *reverseConnectIndicator = (void *)(uintptr_t)0xFFFFFFFF;
+
+    UA_Client *client = (UA_Client*)application;
+
+    UA_LOCK(&client->clientMutex);
+
+    if (!client->channel.connectionId) {
+        client->channel.connectionId = connectionId;
+        *connectionContext = reverseConnectIndicator;
+    }
+
+    if (*connectionContext == reverseConnectIndicator && state == UA_CONNECTIONSTATE_CLOSING) {
+        if ((client->channel.state == UA_SECURECHANNELSTATE_REVERSE_LISTENING || client->channel.state == UA_SECURECHANNELSTATE_CLOSING)
+                && client->channel.connectionId == connectionId) {
+            client->channel.state = UA_SECURECHANNELSTATE_CLOSED;
+            notifyClientState(client);
+        }
+        UA_UNLOCK(&client->clientMutex);
+        return;
+    }
+
+    if (client->channel.connectionId == connectionId &&  *connectionContext == reverseConnectIndicator) {
+        client->channel.state = UA_SECURECHANNELSTATE_REVERSE_LISTENING;
+        notifyClientState(client);
+    }
+
+    /* This is a connection initiated by a server, disconnect the listener */
+    if (client->channel.connectionId != connectionId) {
+        cm->closeConnection(cm, client->channel.connectionId);
+    }
+
+    /* First call with the new connection from the server, reset secure channel information */
+    if (client->channel.connectionId != connectionId && *connectionContext == reverseConnectIndicator) {
```

Comment:
This can be merged with the if-condition just above.
Because the the new active connection will inherit the reverseConnectIndicator.
And we only reset it during the first callback.

Reply:
Done